File: app/models/review.rb
# frozen_string_literal: true

class Review < ApplicationRecord
  belongs_to :job
  belongs_to :reviewer, class_name: "User", foreign_key: "reviewer_id"
  belongs_to :reviewee, class_name: "User", foreign_key: "reviewee_id"

  validates :rating, presence: true, numericality: {in: 1..5}
  validates :comment, length: {maximum: 1000}
  validates :job_id, presence: true
  validates :reviewer_id, presence: true
  validates :reviewee_id, presence: true
  validates :job_id, uniqueness: {scope: [:reviewer_id, :reviewee_id], message: "can only be reviewed once by this user for this reviewee"}

  # Scopes for filtering reviews
  scope :by_job, ->(job_id) { where(job_id: job_id) }
  scope :by_reviewer, ->(reviewer_id) { where(reviewer_id: reviewer_id) }
  scope :by_reviewee, ->(reviewee_id) { where(reviewee_id: reviewee_id) }
  scope :positive, -> { where("rating >= ?", 4) }
  scope :negative, -> { where("rating <= ?", 2) }
  scope :average_rating, -> { average(:rating) }
end

File: app/frontend/types/review.ts
import { User } from './auth';

export interface Review {
  id?: number;
  job_id: number;
  reviewer_id: number;
  reviewee_id: number;
  rating: number;
  comment?: string;
  created_at?: string;
  updated_at?: string;
  reviewer?: User;
  reviewee?: User;
  errors?: Record<string, string[]>;
}

export interface ReviewFormProps {
  job: { id: number; title: string };
  reviewee_id: number; // Corrected to directly include reviewee_id
  errors: Record<string, string[]>;
}
